drivers/volume/portworx/portworx.go

```diff
@@ -744,75 +862,174 @@ func (p *portworx) createVolumeSnapshotData(
 	if err != nil {
 		return nil, err
 	}
 
 	return k8s.Instance().GetPersistentVolumeClaim(parentPV.Spec.ClaimRef.Name, parentPV.Spec.ClaimRef.Namespace)
 }
 
-func (p *portworx) checkCloudSnapStatus(volID string) (string, error) {
+func (p *portworx) waitForCloudSnapCompletion(op, volID string) error {
+	return wait.ExponentialBackoff(cloudsnapBackoff, func() (bool, error) {
+		_, pendingMsg, err := p.checkCloudSnapStatus(op, volID)
+		if err != nil {
+			return true, err
+		}
+
+		if len(pendingMsg) > 0 {
+			return false, nil
+		}
+
+		return true, nil
+	})
+}
+
+func (p *portworx) checkCloudSnapStatus(op, volID string) (string /* cs ID */, string /* pending msg */, error) {
 	response, err := p.volDriver.CloudBackupStatus(&api.CloudBackupStatusRequest{
 		SrcVolumeID: volID,
 	})
 	if err != nil {
-		return "", err
+		return "", "", err
 	}
 
 	csStatus, present := response.Statuses[volID]
 	if !present {
-		return "", fmt.Errorf("failed to get cloudsnap status for volume: %s", volID)
+		return "", "", fmt.Errorf("failed to get cloudsnap status for volume: %s", volID)
 	}
 
 	statusStr := getCloudSnapStatusString(&csStatus)
+
+	if csStatus.Status == cloudSnapStatusFailed {
+		return csStatus.ID, "", fmt.Errorf("cloudsnap %s for %s failed. %s", op, volID, statusStr)
+	}
+
 	if csStatus.CompletedTime.IsZero() {
-		return "", fmt.Errorf("cloudsnap for %s still not completed. %s", volID, statusStr)
+		return csStatus.ID, fmt.Sprintf("cloudsnap %s for %s still not completed. %s", op, volID, statusStr), nil
+
+	}
+
+	if !strings.Contains(csStatus.Status, cloudSnapStatusDone) {
+		return csStatus.ID, fmt.Sprintf("cloudsnap %s for %s still not done. %s", op, volID, statusStr), nil
+	}
+
+	return csStatus.ID, "", nil
+}
+
+// revertPXSnaps deletes all given snapIDs
+func (p *portworx) revertPXSnaps(snapIDs []string) {
+	failedDeletions := make(map[string]error, 0)
+	for _, id := range snapIDs {
+		err := p.volDriver.Delete(id)
+		if err != nil {
+			failedDeletions[id] = err
+		}
+	}
+
+	if len(failedDeletions) > 0 {
+		errString := ""
+		for failedID, failedErr := range failedDeletions {
+			errString = fmt.Sprintf("%s delete of %s failed due to err: %v.\n", errString, failedID, failedErr)
+		}
+
+		logrus.Errorf("failed to revert created PX snapshots. err: %s", errString)
+		return
+	}
+
+	logrus.Infof("successfully reverted PX snapshots")
+	return
+}
+
+func (p *portworx) revertSnapObjs(snapObjs []*crdv1.VolumeSnapshot) {
+	failedDeletions := make(map[string]error, 0)
+
+	for _, snap := range snapObjs {
+		err := wait.ExponentialBackoff(snapAPICallBackoff, func() (bool, error) {
+			deleteErr := k8s.Instance().DeleteSnapshot(snap.Metadata.Name, snap.Metadata.Namespace)
+			if deleteErr != nil {
+				logrus.Infof("failed to delete volumesnapshot due to: %v", deleteErr)
+				return false, nil
+			}
+
+			return true, nil
+		})
+		if err != nil {
+			failedDeletions[fmt.Sprintf("[%s] %s")] = err
```

Comment:
No variables for those specifiers in Sprintf()?

Reply:
sorry, this got checked in incomplete.. fixed